controllers/mirrorpeer_controller.go

```diff
@@ -39,17 +44,54 @@ type MirrorPeerReconciler struct {
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the MirrorPeer object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
 func (r *MirrorPeerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	// Fetch MirrorPeer for given Request
+	var mirrorPeer multiclusterv1alpha1.MirrorPeer
+	err := r.Get(ctx, req.NamespacedName, &mirrorPeer)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			logger.Info("Could not find MirrorPeer. Ignoring since object must have been deleted")
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		logger.Error(err, "Failed to get MirrorPeer")
+		return ctrl.Result{}, err
+	}
 
-	// your logic here
+	// Validate MirrorPeer
+	// MirrorPeer.Spec must be defined
+	if reflect.DeepEqual(mirrorPeer.Spec, multiclusterv1alpha1.MirrorPeerSpec{}) {
+		return ctrl.Result{Requeue: false}, fmt.Errorf("validation: MirrorPeer.Spec must not be empty")
+	}
+	// MirrorPeer.Spec.Items must be unique
+	if reflect.DeepEqual(mirrorPeer.Spec.Items[0], mirrorPeer.Spec.Items[1]) {
+		return ctrl.Result{Requeue: false}, fmt.Errorf("validation: PeerRef must be unique within a MirrorPeer object")
+	}
+	for i := range mirrorPeer.Spec.Items {
+		// MirrorPeer.Spec.Items must not have empty fields
+		if mirrorPeer.Spec.Items[i].ClusterName == "" || mirrorPeer.Spec.Items[i].StorageClusterRef.Name == "" || mirrorPeer.Spec.Items[i].StorageClusterRef.Namespace == "" {
+			// return error and do not requeue since user needs to update the spec
+			// when user updates the spec, new reconcile will be triggered
+			return reconcile.Result{Requeue: false}, fmt.Errorf("validation: PeerRef must not be empty or undefined")
+		}
+		// MirrorPeer.Spec.Items[*].ClusterName must be a valid ManagedCluster
+		var mcluster clusterv1.ManagedCluster
+		err = r.Client.Get(ctx, types.NamespacedName{Name: mirrorPeer.Spec.Items[i].ClusterName}, &mcluster)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return ctrl.Result{Requeue: false}, fmt.Errorf("validation: cluster %q is not a ManagedCluster", mirrorPeer.Spec.Items[i].ClusterName)
+			}
+			return ctrl.Result{}, fmt.Errorf("validation: unable to get ManagedCluster %q", mirrorPeer.Spec.Items[i].ClusterName)
```

Comment:
Print the `err` too?

Reply:
controller-runtime prints it while re-queuing the request.

Reply:
I didn't know that, how?

Reply:
When we return an error and want to requeue, controller runtime logs the reason for requeue and adds it to queue.
I will try to find code reference and get the logs from a running cluster.

Reply:
```bash
2021-08-05T16:20:52.148+0530	ERROR	controller-runtime.manager.controller.mirrorpeer	Reconciler error	{"reconciler group": "multicluster.odf.openshift.io", "reconciler kind": "MirrorPeer", "name": "mirrorpeer-example", "namespace": "", "error": "validation: unable to get ManagedCluster \"cluster-1\""}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/uchapaga/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.0-beta.4/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/home/uchapaga/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.0-beta.4/pkg/internal/controller/controller.go:214

2021-08-05T16:20:52.154+0530	ERROR	controller-runtime.manager.controller.mirrorpeer	Reconciler error	{"reconciler group": "multicluster.odf.openshift.io", "reconciler kind": "MirrorPeer", "name": "mirrorpeer-example", "namespace": "", "error": "validation: unable to get ManagedCluster \"cluster-1\""}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/uchapaga/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.0-beta.4/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/home/uchapaga/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.0-beta.4/pkg/internal/controller/controller.go:214

```

Reply:
Ok but where is `err`? I just see the `fmt.Errorf()` message.

Reply:
🤦🏽 Now I see what you mean. Adding the actual error to the message 😅